Assets/MixedRealityToolkit/Utilities/Physics/TwoHandMoveLogic.cs

```diff
@@ -66,12 +71,45 @@ public Vector3 Update(MixedRealityPose pointerCentroidPose, Quaternion objectRot
                 Vector3 adjustedPointerToGrab = (pointerLocalGrabPoint * distanceRatio);
                 adjustedPointerToGrab = pointerCentroidPose.Rotation * adjustedPointerToGrab;
 
-                return adjustedPointerToGrab - objectRotation * scaledGrabToObject + pointerCentroidPose.Position;
+                newPosition = adjustedPointerToGrab - objectRotation * scaledGrabToObject + pointerCentroidPose.Position;
             }
             else
             {
-                return pointerCentroidPose.Position + pointerToObject;
+                newPosition = pointerCentroidPose.Position + pointerToObject;
+            }
+
+            // If instructed to restrict movement to a specific axis, enforce that after all other translation logic
+            // has run.
+            if (movementConstraint == MovementConstraintType.TranslationAxes)
+            {
+                newPosition = HandleTranslationAxesRestrictions(translationAxes, newPosition);
+            }
+            return newPosition;
+        }
+
+        /// <summary>
+        /// Returns a Vector3 for the new position of the object being manipulated based on
+        /// the new desired location restricted by the allowed axes of freedom.
+        /// </summary>
+        /// <remarks>
+        /// For example, if translationAxes is only AxisPreference.X, this will ensure that
+        /// the object will only move in the X direction.
+        /// </remarks>
+        private Vector3 HandleTranslationAxesRestrictions(AxisPreference translationAxes, Vector3 position)
+        {
+            if (!translationAxes.HasFlag(AxisPreference.X))
```

Comment:
It looks like this is doing constrained movement in world space, meaning that if we set constraint to be X, then the object will only move along the world - space x axis. The tricky thing about world-space X is that this is pretty arbitrary on HoloLens. The world-space +X axis points to the right of wherever the user was when she started the app. So, this X axis will change depending on where she is facing when she starts the app. It also is not related to the actual object's x axis, meaning that if the object is rotated a bit and constraint is X, it may appear to be stuck on a relatively arbitrary axis.

I think the translation constraint should by default be in the object's local space, or at least using the coordinate system that the object had when manipulation started.

Reply:
I second this, when I experimented with this for my WIP constraint changes, local space constraints felt better to me than world space constraints. Perhaps we should provide an option to switch between local and global translation constraints, like you mention [here](https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5147#issuecomment-532430422) for rotation @wiwei.